doc/controller-card_poweroff.md

```diff
@@ -0,0 +1,143 @@
+/components/component[name=controller0]/controller-card/state/power-admin-state, POWER_ENABLED
+/components/component[name=controller1]/state/redundant-role, SECONDARY
+/components/component[name=controller1]/controller-card/state/power-admin-state, POWER_ENABLED
+```
+
+After controller-card0 transitions to redundant-role SECONDARY:
+
+```
+/components/component[name=controller0]/state/redundant-role, SECONDARY
+/components/component[name=controller0]/controller-card/config/power-admin-state, POWER_DISABLED
+/components/component[name=controller0]/controller-card/state/power-admin-state, POWER_DISABLED
+/components/component[name=controller0]/state/last-poweroff-reason/trigger, USER_INITIATED
+/components/component[name=controller0]/state/last-poweroff-reason/details, "User initiated Shutdown"
+/components/component[name=controller0]/state/last-poweroff-time, 1706590714681765937
+/components/component[name=controller1]/state/redundant-role, PRIMARY
+/components/component[name=controller1]/controller-card/config/power-admin-state, POWER_ENABLED
+/components/component[name=controller1]/controller-card/state/power-admin-state, POWER_ENABLED
+```
+
+2. A controller-card which is in state/redundant-role=SECONDARY and is config/power-admin-state=POWER_DISABLED should remain powered off, even after a reboot. It is possible that after a reboot, both the controller cards are powered-on. However as soon as the configuration is loaded, the system should power-off the subject controller-card.
+
+3. A CONTROLLER_CARD in redundant-role SECONDARY and state/power-admin-state = POWER_DISABLED cannot transition to redundant-role PRIMARY. If the PRIMARY CONTROLLER_CARD goes down, the device will be offline.
+
+4. When a controller card boots up and loads it’s configuration, power-admin-state may be set to POWER_DISABLED. The CONTROLLER_CARD should then power off and never enter into any controller card primary/secondary election process. This also means that an implementation shouldn't start a controller card election process until the configuration is loaded and consumed.
+
+5. If the chassis has a single controller card and it was configured for config/power-admin-state=POWER_DISABLED, as per Rule#1 above, its state/power-admin-state will stay as POWER_ENABLED given that its state/redundant-role=PRIMARY. However during the next reboot, if the chassis continues to have only one controller-card then it must ignore the configuration of config/power-admin-state=POWER_DISABLED and continue to bootup as PRIMARY. In this situation though, the implementation must send a Syslog message of the severity "Warning" to inform the Operator about the situation. Post reboot, the config/power-admin-state=POWER_DISABLED will stay as is and state/power-admin-state will continue to show as POWER_ENABLED. In this case, state/last-poweroff-time, state/last-poweroff-reason/trigger and state/last-poweroff-reason/details must not be updated
```

Comment:
This means that the faulty card cannot be disabled if you encounter a failure in communication between the primary and the secondary cards (persistent split brain scenario)

If this is acceptable, I think that should be explicitly mentioned in the text.

Reply:
I didn't ask to add all hypothetical situations :-)

There are a few challenges that I see, and why I asked about this specifically.

In one of the stated objectives, you say

> Operator suspects a network device issue is occurring due to faults on a CONTROLLER_CARD. The operator wants to quickly disable the CONTROLLER_CARD and later, troubleshoot the issue in a maintenance window. The operator also wants to ensure that the card stays disabled even if the device is rebooted.

So, let's say, an operator has a problem where split brain scenario has occurred - he definitely will want to disable the failed card. But now he cannot truly disable it, if the failed card is not able to see the other card, and thinks that it's primary. Because it follows the rules specified in this document (the only card is not disabled immediately, and post reboot has to check if it's the only card and boot successfully if it is)

Thus, I think it is important to call out that the rules specified here explicitly disallow using the `POWER_DISABLED` configuration to mitigate the split-brain issue if it occurs.

And no, I don't agree that
 > Possibility of such a command to power-off a controller-card using configuration do not change how the implementation should otherwise deal with operational challenges like a Split-brain scenario.

This document
a) enforces a particular architecture of NOS boot process, and active-standby selection (which, in my opinion, should not be standardized by OpenConfig)
b) says that power-off rules cannot be used to mitigate the split brain in certain conditions, where another set of rules w.r.t a single card scenario behavior may produce a completely different result.

Reply:
`So, let's say, an operator has a problem where split brain scenario has occurred - he definitely will want to disable the failed card. But now he cannot truly disable it, if the failed card is not able to see the other card, and thinks that it's primary.`

***Response:** 
- At first, we are also assuming that the operator is able to use push config to the box in a split brain scenario.

`Thus, I think it is important to call out that the rules specified here explicitly disallow using the POWER_DISABLED configuration to mitigate the split-brain issue if it occurs.`

***Response***
I strongly feel that is implicit in the rule already, a Primary controller-card will not be turned off during run time (in a split-brain scenario both cards tend to be PRIMARY and hence none can be turned-off if somehow the operator manages to configure one). Also, if it deems itself as the lone controller-card during reboot, it will continue with the boot process (therefore in the split-brain scenario both cards will continue to boot. However the NOS is expected to have proper safety in place to prevent such a situation at boot). Calling out specific operational failure example in the rules will be repetitive and unproductive. But, this is covered in the `Concerns and possible failure scenarios` section below.

a) enforces a particular architecture of NOS boot process, and active-standby selection (which, in my opinion, should not be standardized by OpenConfig)

**Response**
For this process to work as expected, it is important that the implementation learns what is in the config prior to starting the election.

`b) says that power-off rules cannot be used to mitigate the split brain in certain conditions, where another set of rules w.r.t a single card scenario behavior may produce a completely different result.`

**Response**
Right, that is the expectation and I dont see any ambiguity there. Attached flowchart must help as well. Reiterating the rules in context
- Primary controller card cannot be powered-off, only the secondary-controller card can be powered-off during runtime.  During Split-brain situation when both cards are Primary at run-time, and somehow if the operator can configure one card as disabled, it will stay up because it deems itself as PRIMARY.
- During the boot process, if the controller-card is the only card, then the POWER_DISABLED configuration must be ignored and the boot process should continue. Now if during the boot process there is a split brain situation then the implementation MUST have proper controls in place for protection. Introduction of config/power-admin-state = POWER_DISABLED/POWER_ENABLED is orthogonal to such an expectation for protection from the implementation.